mdn/src/main/java/org/apache/james/mdn/MDNReport.java

```diff
@@ -0,0 +1,135 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mdn;
+
+import java.util.Optional;
+
+import org.apache.james.mdn.fields.Disposition;
+import org.apache.james.mdn.fields.FinalRecipient;
+import org.apache.james.mdn.fields.OriginalMessageId;
+import org.apache.james.mdn.fields.OriginalRecipient;
+import org.apache.james.mdn.fields.ReportingUserAgent;
+
+import com.google.common.base.Preconditions;
+
+public class MDNReport {
+
+    public static class Builder {
+        private Optional<ReportingUserAgent> reportingUserAgentField = Optional.empty();
+        private Optional<OriginalRecipient> originalRecipientField = Optional.empty();
+        private Optional<FinalRecipient> finalRecipientField = Optional.empty();
+        private Optional<OriginalMessageId> originalMessageIdField = Optional.empty();
+        private Optional<Disposition> dispositionField = Optional.empty();
+
+        public Builder reportingUserAgentField(ReportingUserAgent reportingUserAgentField) {
+            this.reportingUserAgentField = Optional.of(reportingUserAgentField);
+            return this;
+        }
+
+        public Builder originalRecipientField(OriginalRecipient originalRecipientField) {
+            this.originalRecipientField = Optional.of(originalRecipientField);
+            return this;
+        }
+
+        public Builder originalRecipientField(Optional<OriginalRecipient> originalRecipientField) {
+            this.originalRecipientField = originalRecipientField;
+            return this;
+        }
+
+        public Builder finalRecipientField(FinalRecipient finalRecipientField) {
+            this.finalRecipientField = Optional.of(finalRecipientField);
+            return this;
+        }
+
+        public Builder originalMessageIdField(OriginalMessageId originalMessageIdField) {
+            this.originalMessageIdField = Optional.of(originalMessageIdField);
+            return this;
+        }
+
+        public Builder dispositionField(Disposition dispositionField) {
+            this.dispositionField = Optional.of(dispositionField);
+            return this;
+        }
+
+        public MDNReport build() {
+            Preconditions.checkState(reportingUserAgentField.isPresent());
```

Comment:
what is the point of using optional if it's not optional ?

Reply:
1. It becomes optional later in the PR
2. I really don't like null values.